### PR TITLE
Allow user to specify tests we want jtreg to run

### PR DESCRIPTION
**Pull request by Bluevisitor**
Currently we allow the user to say "run the default jtreg tests after the build is complete", but we don't allow the user to specify a subset.

If we want to run less tests than the default, or more, or if we want to run focused testing (in the event that a specific module has been changed recently), we'll need this option or something like it.

### CODE REVIEW

**Comment by Bluevisitor** [comment]
The test run didn't work: https://ci.adoptopenjdk.net/job/adam_test/8/consoleFull

My local run continues to work.

Will investigate.

File: makejdk.sh [hunk 1]
```diff
@@ -215,6 +218,10 @@ else
   $WORKING_DIR/sbin/build.sh $WORKING_DIR $TARGET_DIR $OPENJDK_REPO_NAME $BUILD_FULL_NAME $JVM_VARIANT
 
   if [[ ! -z $JTREG ]]; then
-    $WORKING_DIR/sbin/jtreg.sh $WORKING_DIR $OPENJDK_REPO_NAME $BUILD_FULL_NAME
+    if [[ ! -z $JTREG_TEST_SUBSETS ]]; then
+      $WORKING_DIR/sbin/jtreg.sh $WORKING_DIR $OPENJDK_REPO_NAME $BUILD_FULL_NAME $JTREG_TEST_SUBSETS
```
Comment (karianna) [inline makejdk.sh]: Could we default subsets to be "all" or empty string or whatever indicates all?

**Comment by Bluevisitor** [comment]
We're missing a semicolon from the non-local version. Doh.

File: makejdk.sh [hunk 2]
```diff
@@ -75,6 +75,9 @@ while [[ $# -gt 0 ]] && [[ ."$1" = .-* ]] ; do
 
     "--jtreg" | "-j" )
     JTREG=true; shift;;
+    
+    "--jtreg_subsets" )
+    JTREG=true; JTREG_TEST_SUBSETS="$1" shift;;
```
Comment (karianna) [inline makejdk.sh]: See comment below, if --jtreg runs with a default of all, wouldn't need this separate option

File: sbin/jtreg.sh [hunk 3]
```diff
@@ -91,7 +92,11 @@ settingUpEnvironmentVariablesForJTREG()
 runJtregViaMakeCommand()
 {
   echo "Running jtreg via make command (debug logs enabled)"
-  make test jobs=10 LOG=debug  
+  if [ -z $JTREG_TEST_SUBSETS }; then
```
Comment (karianna) [inline sbin/jtreg.sh]: See comment above, is it possible to default to all

File: makejdk.sh [hunk 4]
```diff
@@ -75,6 +75,9 @@ while [[ $# -gt 0 ]] && [[ ."$1" = .-* ]] ; do
 
     "--jtreg" | "-j" )
     JTREG=true; shift;;
+    
+    "--jtreg_subsets" )
+    JTREG=true; JTREG_TEST_SUBSETS="$1"; shift;;
```
Comment (karianna) [inline makejdk.sh]: Could we default subsets to be "all" or empty string or whatever indicates all?

File: makejdk.sh (hunk 1 above)
Comment (Bluevisitor) [inline makejdk.sh]: Currently, when we run the jtreg without options, we default to jdk_core and langtools (the latter doesn't seem to work, but jdk_core has enough coverage that we don't mind atm).

We could default to jdk_core, which gives us a core set of tests and removes the non-functional langtools, or we could switch to jdk_all. I don't know how long jdk_all takes, but I'm guessing it's longer than jdk_core. 

The latter already takes 25 minutes on our ultra-fast linux server.

File: sbin/jtreg.sh (hunk 3 above)
Comment (Bluevisitor) [inline sbin/jtreg.sh]: Currently, when we run the jtreg without options, we default to jdk_core and langtools (the latter doesn't seem to work, but jdk_core has enough coverage that we don't mind atm).

We could default to jdk_core, which gives us a core set of tests and removes the non-functional langtools, or we could switch to jdk_all. I don't know how long jdk_all takes, but I'm guessing it's longer than jdk_core.

The latter already takes 25 minutes on our ultra-fast linux server.

File: makejdk.sh (hunk 4 above)
Comment (Bluevisitor) [inline makejdk.sh]: Currently, when we run the jtreg without options, we default to jdk_core and langtools (the latter doesn't seem to work, but jdk_core has enough coverage that we don't mind atm).

We could default to jdk_core, which gives us a core set of tests and removes the non-functional langtools, or we could switch to jdk_all. I don't know how long jdk_all takes, but I'm guessing it's longer than jdk_core.

The latter already takes 25 minutes on our ultra-fast linux server.

File: makejdk.sh (hunk 1 above)
Comment (karianna) [inline makejdk.sh]: Do we have a separate issue tracking why langtools is failing?

Ideally we'd default to jdk_all for release builds and yeah probably jdk_core for nightly builds.  So perhaps this script should default to jdk_core (a decent speed/coverage) compromise, but we can override that in our various build scripts.

IIRC jdk_all will also perform UI tests (Swing/AWT) - so will expose other missing deps that our build boxes may need.

Comment (Bluevisitor) [inline makejdk.sh]: I'm not aware of any issue tracking why langtools is failing. The error is:

/usr/bin/make -k -C  ../langtools/test  JT_JAVA=/home/jenkins/workspace/openjdk_build_x86-64_linux/openjdk/build/linux-x86_64-normal-server-release/images/j2sdk-image JTREG_HOME= TEST=jtreg jtreg
make[2]: Entering directory '/home/jenkins/workspace/openjdk_build_x86-64_linux/openjdk/langtools/test'
make[2]: *** No rule to make target '/bin/jtreg', needed by 'check-jtreg'.

If we specify jdk_core specifically, langtools won't be run. If we don't specify any test, then jck_core and langtools will be run by default.

The question is: do we still want to be running langtools?

Perhaps we should have an issue for langtools, and also specify jdk_core in the jenkins setup by default?

Comment (karianna) [inline makejdk.sh]: Ok, I'm adding that failure to the overall test issue #7 - I think we probably do want to be running langtools, I'll let @neomatrix369 comment over on #7

**Comment by Bluevisitor** [comment]
The test run passed: https://ci.adoptopenjdk.net/job/adam_test/10

I see two problems here.

1) The .jtr.xml files we find the results xml in are not in the jdk_core folder if we're running alternate tests (like jdk_lang). We can resolve this by replacing "jdk_core" with a wildcard character in the post-build "Junit xml interpretation" stage.

2) The results files are also not in jdk_core, so our test harness is having a tough time uploading it. I plan to tweak the relevant scripts so they can accommodate this.

**Comment by Bluevisitor** [comment]
Ok, I think we're all good.

Note: The RUN_JTREG settings (including default value and comment) needs to be copied to the main linux jenkins job once this change has been merged. Also, the "jdk_core" in the "location of results xml files" post-build step needs to be replaced with a "*" (sans quotes).

**Comment by Bluevisitor** [comment]
Here's a test run to verify the changes.

https://ci.adoptopenjdk.net/view/all/job/adam_test/11/

File: makejdk.sh (hunk 4 above)
Comment (karianna) [inline makejdk.sh]: Should JTREG_TEST_SUBSETS default to 'the default *before* having $1 (if $1 exists / is non empty) assigned to it?  That way you may not need to have the separate --jtreg_subsets flag?

File: makejdk.sh (hunk 1 above)
Comment (karianna) [inline makejdk.sh]: Hoping that ` if [[ ! -z $JTREG_TEST_SUBSETS ]]; then` won't be required...

File: makejdk.sh (hunk 4 above)
Comment (Bluevisitor) [inline makejdk.sh]: After our call I understand what you mean. Will make those changes now.

Comment (Bluevisitor) [inline makejdk.sh]: After playing with the shell script, I've determined that shifting after assigning $1 to it has the nasty side-effect of swallowing the next variable passed in on the command line.

so "./makejdk.sh --jtreg --potato" assigns "--potato" to JTREG_TEST_SUBSETS, and removes --potato from the list of arguments to be processed.

If we don't want to add another "if", what we could do is require a value for jtreg.

'./makejdk.sh --jtreg "" --potato' works, for example.

Will make that change. Let me know what you think when you get back.

Comment (Bluevisitor) [inline makejdk.sh]: karianna suggests that we're making a mountain out of a molehill here, so he proposed (and I agreed) that we should add the functionality now and tidy it up later when we have spare time.

**Comment by Bluevisitor** [comment]
Thanks. :)

**Pull request merged**